src/Oscoin/Protocol.hs

```diff
@@ -0,0 +1,287 @@
+module Oscoin.Protocol
+    ( Protocol -- opaque
+    , Handle -- opaque
+    , ShutdownPolicy(..)
+
+    -- * Getters
+    , protoOrphanage
+
+    -- * Running and using the Protocol
+    , withProtocol
+    , runOscoinProtocol
+
+    -- * Internals
+    , stepProtocol
+    ) where
+
+import           Oscoin.Prelude
+
+import           Formatting (Buildable, Format, shown, (%))
+import           Oscoin.Consensus (validateBlockchain)
+import           Oscoin.Consensus.Config as Consensus
+import           Oscoin.Consensus.Types (Validate)
+import           Oscoin.Crypto.Blockchain (Blockchain(..))
+import           Oscoin.Crypto.Blockchain.Block
+import           Oscoin.Crypto.Hash (HasHashing, Hash)
+import           Oscoin.Storage.Block.Abstract
+                 ( BlockStore
+                 , BlockStoreInternals
+                 , FullBlockStoreAPI
+                 , GetBlocksQuery(..)
+                 )
+import qualified Oscoin.Storage.Block.Abstract as BlockStore
+import           Oscoin.Storage.Block.Orphanage (Orphanage)
+import qualified Oscoin.Storage.Block.Orphanage as O
+import           Oscoin.Telemetry (fmtBlockHash, fmtValidationError)
+import           Oscoin.Telemetry.Logging (LogMsg, ftag)
+import           Oscoin.Time.Chrono as Chrono
+
+import           Control.Concurrent.Async (Async)
+import qualified Control.Concurrent.Async as Async
+import           Control.Concurrent.STM
+import qualified Data.List.NonEmpty as NonEmpty
+
+-- | A stateful handle over a 'Protocol'.
+data Handle c tx s = Handle
+    { hProtocol :: MVar (Protocol c tx s IO) }
```

Comment:
My intention was to have exclusive access to the blockStore when processing a block, so that my calls to `extendsTip`, `getTip` weren't influenced by other things happening concurrently. 

When taking another look now, though, I do appreciate this feels redundant, especially in the presence of the blockstore API split. If anything, the `Protocol` is supposed to be "one and only" for the entire lifecycle of the application, so it's virtually the only component which write stuff into the blockstore.

Reply:
We discussed this together and we agreed that indeed `stepProtocol` is a critical section as it performs the delicate task of extending the chain, and as such it makes sense to run in a mutex. (5fa7b38).